### PR TITLE
Redesign Persistent Batch in vLLM

### LINKED ISSUE #23446: [RFC]: Redesigning Persistent Batch in vLLM
### Motivation.

To reduce the overhead of input data construction, vLLM v1 introduced the **persistent batch** technique, inspired by [[LMDeploy](https://github.com/InternLM/lmdeploy)](https://github.com/InternLM/lmdeploy).

## How vLLM's Persistent Batch Works

At a high level, the mechanism operates as follows:

1. Each request’s state is stored in an `InputBatch`, which consists of CPU tensors shaped either as `[num_reqs]` (e.g., sampling temperature) or `[num_reqs, L]` (e.g., block tables).
2. When `N` requests are scheduled, we reorder `InputBatch` so that the first `N` entries correspond to the scheduled requests.
3. Finally, these `N` entries are then copied from CPU to GPU (using regular `cudaMemcpyAsync`).

The key insight is that the set of scheduled requests typically changes only slightly between steps.
Thus, the `InputBatch` from the previous step can be reused with only incremental updates (e.g., appending `new_block_ids` to the block table and reordering a few rows when some requests finish).

## Limitations of Current Implementation

While effective in many cases, vLLM's persistent batch has several drawbacks:

1. **Efficiency depends on minimal changes between steps**—a condition that doesn’t always hold:
   * (while it's not the default case in vLLM) A scheduler may alternate between prefill and decode batches.
   * With pipeline parallelism, each (micro)batch can involve a completely different set of requests, forcing frequent `InputBatch` flushes every step.

2. **Duplicate state storage**: A secondary structure (`CachedRequestState`) is required to hold states for requests that _are unfinished but not scheduled in the current step_.
   * This introduces inefficiency since two copies of the same states must be updated each step.
   * It also complicates the code (two sources of truth).

3. **Complex request reordering**: Some attention implementations require grouping prefill and decode requests separately. To meet this constraint, we need to re-order all the persistent batch, which is complex and error-prone.

### Proposed Change.

To address these issues, I propose a new design based on separating **persistent request states** from **step-level input batches**.

Image

### 1. Persistent Request States

We introduce a `RequestStates` class, which holds all persistent request metadata. Most states are stored as (UVA-mapped) **CPU** tensors, with one entry (or row) per request.

```python
class RequestStates:
    def __init__(self, ...):
        self.req_id_to_idx: dict[str, int] = {}
        self.req_idx_to_id: dict[int, str] = {}
        max_cached_reqs = 2 * max_num_reqs
        self.temperature = torch.zeros(max_cached_reqs, device="cpu", pin_memory=True)
        ...
```

Key differences from the old approach:

* **No reordering**: Each request is assigned a fixed index when first scheduled. This index does not change until the request completes.
* **Single source of truth**: `RequestStates` replaces `CachedRequestState`, eliminating duplicated state storage.

---

### 2. Step-Level Input Batches

On each step, a new `InputBatch` is constructed from `RequestStates`.

```python
@dataclass
class InputBatch:
    idx_mapping: torch.Tensor   # batch_idx -> request_idx
    block_table: torch.Tensor
    slot_mapping: torch.Tensor
    sampling_metadata: SamplingMetadata
    ...
```

`InputBatch` contains the **dense GPU tensors** actually fed into the model and sampler.
**Importantly, the construction of `InputBatch` is fast because it happens on GPU**. Given `index_mapping` (determined by the attention backend), we write a custom Triton kernel that gathers entries from `RequestStates` into dense GPU tensors. The kernel can access the CPU tensors in `RequestStates` because they are using UVA.

---

### 3. Special Handling for Block Tables

Image

Block tables are treated differently to minimize CPU–GPU transfers:

* Their persistent state is kept on GPU (unlike other states).
* Each step, `new_block_ids` from the scheduler are incrementally applied directly on GPU.
* A custom Triton kernel gathers the final block table for input.
* `slot_mapping` is also built on GPU using the updated block table.

---

## Benefits of the Redesign

* **Simplicity**: Eliminates reordering logic, reducing complexity and sources of bugs.
* **Performance**:
  * Fewer CPU updates to persistent states (no reordering).
  * Reduced CPU–GPU communication by keeping block tables on GPU.
  * Faster input construction via GPU kernels (e.g., slot mapping).

---

Prototype(WIP): https://github.com/vllm-project/vllm/tree/woosuk/input-prep

### Feedback Period.

_No response_

### CC List.

cc @njhill @LucasWilkinson @zhuohan123 

### Any Other Things.

_No response_

### Before submitting a new issue...

- [x] Make sure you already searched for relevant issues, and asked the chatbot living at the bottom right corner of the [documentation page](https://docs.vllm.ai/en/latest/), which can answer lots of frequently asked questions.

### PR DESCRIPTION
**Pull request by WoosukKwon**
Closes #23446

### CODE REVIEW

File: vllm/v1/worker/gpu_model_runner.py [hunk 1]
```diff
@@ -2165,20 +1997,19 @@ def _get_prompt_logprobs_dict(
...
 
             num_nans_in_logits = {}
             num_nans_for_index = logits.isnan().sum(dim=-1).cpu().numpy()
-            for req_id in self.input_batch.req_ids:
-                req_index = self.input_batch.req_id_to_index[req_id]
-                num_nans_in_logits[req_id] = (
-                    int(num_nans_for_index[req_index])
-                    if num_nans_for_index is not None
-                    and req_index < logits.shape[0] else 0)
+            for i, req_id in input_batch.req_ids:
```
[the remark this answers is not in the document]
Comment (WoosukKwon) [reply vllm/v1/worker/gpu_model_runner.py]: Fixed.

File: vllm/v1/worker/gpu_model_runner.py [hunk 2]
```diff
@@ -656,84 +547,55 @@ def _get_cumsum_and_arange(
...
         total_num_scheduled_tokens = scheduler_output.total_num_scheduled_tokens
         assert total_num_scheduled_tokens > 0
-        num_reqs = self.input_batch.num_reqs
-        assert num_reqs > 0
+        num_reqs = len(scheduler_output.num_scheduled_tokens)
+
+        # batch_idx -> req_id
+        req_ids = sorted(scheduler_output.num_scheduled_tokens,
+                         key=scheduler_output.num_scheduled_tokens.get)
+
```
Comment (LucasWilkinson) [inline vllm/v1/worker/gpu_model_runner.py]: I assume the attention based reordering will be determined here?

File: vllm/v1/worker/gpu_model_runner.py [hunk 3]
```diff
@@ -656,84 +547,55 @@ def _get_cumsum_and_arange(
...
+        # req_id -> batch_idx
+        req_id_to_batch_idx = {req_id: i for i, req_id in enumerate(req_ids)}
+
+        # batch_idx -> req_idx
+        idx_mapping_list = [
```
Comment (LucasWilkinson) [inline vllm/v1/worker/gpu_model_runner.py:625]: nit: why not just `batch_idx_to_req_state_idx` to be consistent with `req_id_to_batch_idx`

File: vllm/v1/worker/gpu_model_runner.py [hunk 4]
```diff
@@ -828,14 +690,8 @@ def _prepare_inputs(
...
-                blk_table = self.input_batch.block_table[kv_cache_group_id]
-                blk_table_tensor = blk_table.get_device_tensor()[:num_reqs]
-                slot_mapping = blk_table.slot_mapping[:
-                                                      total_num_scheduled_tokens]
-
-                # Fill unused with -1. Needed for reshape_and_cache in full cuda
-                # graph mode.
-                blk_table.slot_mapping[total_num_scheduled_tokens:].fill_(-1)
+                blk_table_tensor = block_tables[kv_cache_group_id]
+                slot_mapping = slot_mappings[kv_cache_group_id]
```
Comment (LucasWilkinson) [inline vllm/v1/worker/gpu_model_runner.py:743]: nice! so much cleaner! 😄

File: vllm/v1/worker/gpu_input_batch.py [hunk 5]
```diff
@@ -1,806 +1,93 @@
...
+    req_id_to_batch_idx: dict[str, int]
+
+    # batch_idx -> req_state_idx
+    idx_mapping: torch.Tensor
+    idx_mapping_np: np.ndarray
+
+    # batch_idx -> num_scheduled_tokens
+    num_scheduled_tokens: np.ndarray
+    total_num_tokens: int
+    max_num_tokens: int
```
Comment (LucasWilkinson) [inline vllm/v1/worker/gpu_input_batch.py]: nit: I think `max_num_tokens` is a bit ambiguous; I think `max_num_tokens_per_request` might be clearer

File: vllm/v1/worker/gpu_model_runner.py [hunk 6]
```diff
@@ -275,13 +259,30 @@ def __init__(
         # None in the first PP rank. The rest are set after load_model.
         self.intermediate_tensors: Optional[IntermediateTensors] = None
 
+        self.block_tables = BlockTables(
+            block_sizes=[self.cache_config.block_size],
+            max_num_reqs=self.max_num_reqs,
+            max_num_cached_reqs=2 * self.max_num_reqs,
```
Comment (LucasWilkinson) [inline vllm/v1/worker/gpu_model_runner.py]: shouldn't this be `max_concurrent_batches * self.max_num_reqs` since `max_concurrent_batches` can be greater than 2 for PP?

Comment (LucasWilkinson) [reply vllm/v1/worker/gpu_model_runner.py]: Do we need to account for preempted requests here too? since I think the cached block-table index should map there request-state index and as far as I can tell the `RequestState` for preempted requests is kept around still occupying an index? Please correct me if this is wrong

Comment (LucasWilkinson) [reply vllm/v1/worker/gpu_model_runner.py]: (ditto comments for 

```
RequestState(
    ...
    max_num_cached_reqs=...
    ...
)
```
)

**Review by LucasWilkinson — COMMENTED** [review]
Im really liking this direction! Did a first pass; will try to do another more in-depth one tmrw.

File: vllm/v1/worker/gpu_input_batch.py (hunk 5 above)
Comment (njhill) [reply vllm/v1/worker/gpu_input_batch.py]: or even `max_tokens_per_request` :)

File: vllm/v1/sample/metadata.py [hunk 7]
```diff
@@ -12,25 +12,24 @@
...
     max_num_logprobs: Optional[int]
 
     no_penalties: bool
-    prompt_token_ids: Optional[torch.Tensor]
     frequency_penalties: torch.Tensor
     presence_penalties: torch.Tensor
     repetition_penalties: torch.Tensor
 
-    output_token_ids: list[list[int]]
+    token_ids: Optional[torch.Tensor]
```
Comment (njhill) [inline vllm/v1/sample/metadata.py:32]: We may still need to track the prompt and output ids separately since there are logits processor impls which need to differentiate (even some of the built-in penalties).

**Review by njhill — COMMENTED** [review]
Looks great to me @WoosukKwon.

I've been thinking about how we can further abstract the sampling metadata handling - I think we can parameterize the triton kernel for that. So that it takes something like `*src_tensors, *dest_tensors, NUM_TENSORS: tl.constexpr`.

And we need to look at how we will handle arbitrary logits processors with other data-structures that need updating based on the index mapping.

For example the min_tokens processor right now applies a sparse update based on specific coordinates within the logits tensor: [here](https://github.com/vllm-project/vllm/blob/98aa16ff41353e3e6c8a3c2f4e933a888dbce1cb/vllm/v1/sample/logits_processor/builtin.py#L296) .. where the slice is a (tensor, tensor) tuple with pair-wise coords of individual elements that should be set.

**Comment by WoosukKwon** [comment]
> I've been thinking about how we can further abstract the sampling metadata handling - I think we can parameterize the triton kernel for that. So that it takes something like *src_tensors, *dest_tensors, NUM_TENSORS: tl.constexpr.

I originally had a similar idea but the problem was that sampling params may have different dtypes, and we don't have a good way to tell their dtypes to Triton.

**Comment by njhill** [comment]
> > I've been thinking about how we can further abstract the sampling metadata handling - I think we can parameterize the triton kernel for that. So that it takes something like *src_tensors, *dest_tensors, NUM_TENSORS: tl.constexpr.
> 
> I originally had a similar idea but the problem was that sampling params may have different dtypes, and we don't have a good way to tell their dtypes to Triton.

In practice I think we won't have many different dtypes though? Could pass as separate lists?
```python
def make_multi_dtype_permute_kernel(
    num_float32_pairs: int = 0,
    num_int32_pairs: int = 0,
):

    @triton.jit
    def multi_dtype_permute_kernel(
        batch_idx_to_req_idx,    # [batch_size]

        # src1, dst1, src2, dst2, ...
        *float32_tensors,
        *int32_tensors,
    ):
        batch_idx = tl.program_id(0)
        req_idx = tl.load(batch_idx_to_req_idx + batch_idx)

        # Process float32 tensors
        for i in range(num_float32_pairs):
            src_tensor = float32_tensors[i * 2]
            dst_tensor = float32_tensors[i * 2 + 1]
            value = tl.load(src_tensor + req_idx, dtype=tl.float32)
            tl.store(dst_tensor + batch_idx, value)

        # Process int32 tensors
        for i in range(num_int32_pairs):
            src_tensor = int32_tensors[i * 2]
            dst_tensor = int32_tensors[i * 2 + 1]
            value = tl.load(src_tensor + req_idx, dtype=tl.int32) 
            tl.store(dst_tensor + batch_idx, value)

    return multi_dtype_permute_kernel
 ```

**Comment by WoosukKwon** [comment]
@njhill Yeah that's not impossible but I'm not sure if it's worth. I will probably not do it in this PR.
